addons/sale/models/sale_order_line.py

```diff
@@ -745,42 +771,49 @@ def _get_protected_fields(self):
+                continue
 
-        price, rule_id = self.order_id.pricelist_id.with_context(product_context).get_product_price_rule(self.product_id, self.product_uom_qty or 1.0, self.order_id.partner_id)
-        new_list_price, currency = self.with_context(product_context)._get_real_price_currency(product, rule_id, self.product_uom_qty, self.product_uom, self.order_id.pricelist_id.id)
-
-        if new_list_price != 0:
-            if self.order_id.pricelist_id.currency_id != currency:
-                # we need new_list_price in the same currency as price, which is in the SO's pricelist's currency
-                new_list_price = currency._convert(
-                    new_list_price, self.order_id.pricelist_id.currency_id,
-                    self.order_id.company_id or self.env.company, self.order_id.date_order or fields.Date.today())
-            discount = (new_list_price - price) / new_list_price * 100
-            if (discount > 0 and new_list_price > 0) or (discount < 0 and new_list_price < 0):
-                self.discount = discount
+            line.discount = 0.0
+            product = line.product_id.with_context(
+                lang=line.order_id.partner_id.lang,
+                partner=line.order_id.partner_id,
+                quantity=line.product_uom_qty,
+                date=line.order_id.date_order,
+                pricelist=line.order_id.pricelist_id.id,
+                uom=line.product_uom.id,
+                fiscal_position=self.env.context.get('fiscal_position'))
+
+            product_context = dict(
+                self.env.context,
+                partner_id=line.order_id.partner_id.id,
+                date=line.order_id.date_order,
+                uom=line.product_uom.id)
+
+            price, rule_id = line.order_id.pricelist_id.with_context(product_context).get_product_price_rule(
+                line.product_id,
+                line.product_uom_qty or 1.0,
+                line.order_id.partner_id)
+            new_list_price, currency = line.with_context(product_context)._get_real_price_currency(
+                product, rule_id, line.product_uom_qty,
+                line.product_uom, line.order_id.pricelist_id.id)
+
+            if new_list_price != 0:
+                if line.order_id.pricelist_id.currency_id != currency:
+                    # we need new_list_price in the same currency as price, which is in the SO's pricelist's currency
+                    new_list_price = currency._convert(
+                        new_list_price, line.order_id.pricelist_id.currency_id,
+                        line.order_id.company_id or self.env.company,
+                        line.order_id.date_order or fields.Date.today())
+                discount = (new_list_price - price) / new_list_price * 100
+                if (discount > 0 and new_list_price > 0) or (discount < 0 and new_list_price < 0):
+                    line.discount = discount
```

Comment:
I find it strange to have this business logic here. Shouldn't it be a method on pricelist or product that returns both the price and the discount? That way we avoid duplication of code (e.g. there is probably a similar code in eCommerce)
We could also have only one compute field that computes both price & discount.

Reply:
This will be done in future task(s), the diff is big enough to begin other structural changes.

Next step is to remove & simplify the ugly context logic used in the prices computation (#79605), but I've a lot of cleanup tasks in mind for the future months.

We'll try to do it progressively, with clear tasks and targets.  PO's and devs are slowly re-discovering the `sale` scope, we want to avoid very big PR's that nobody dares to merge (like the ones I did last year, that we are currently reusing, piece by piece).

Reply:
@fpodoo This is the next step, after this one is merged. :1st_place_medal: